File: script/compile.rs
#!/usr/bin/env rustx

use std::fs;
use std::env;
use std::path;

fn main() {
  let PRO = "cgraph";
  let DIR = path::Path::new(".");
  let INC = DIR.join("include");
  let SRC = DIR.join("src");
  let TST = DIR.join("test");
  let LIB = DIR.join("lib");

  let CC = "cc";
  let mut CFLAGS = "-std=c89 -Wall -pedantic -fPIC".to_string();
  let CSFLAGS = "-shared";
  
  let MODE = "debug";
  if MODE == "debug" {
    CFLAGS.push_str(" -g -DDEBUG");
  } else if MODE == "release" {
    CFLAGS.push_str(" -static -O2");
  }
  
  //  package shared library
  let AR = "ar";
  let ARFLAGS = "-rcs";
  
  for file in SRC.read_dir().expect("read_dir call failed") {
    if let Ok(file) = file {
      if file.path().extension() != None && "c" == file.path().extension().unwrap() {
        let path = file.path();
        let obj = path.with_extension("o");
        println!("{:?} {:?}", path, obj);
      }
    }
  }

  let args: Vec<String> = env::args().collect();
  if args.len() == 1 {
    if !LIB.exists() {
      fs::create_dir(LIB).unwrap();
    }

    for file in SRC.read_dir().expect("read_dir call failed") {
      if let Ok(file) = file {
        if file.path().extension() != None && "c" == file.path().extension().unwrap() {
          let path = file.path();
          let obj = path.with_extension("o");
          println!("{:?} {:?}", path, obj);
        }
      }
    }

  } else if args[1] == "test" {

  } else if args[1] == "clean" {

  } else if args[1] == "distclean" {

  } else if args[1] == "help" {
    println!("{} <target>", file!());
    println!("<target>: ");
    println!("                    compile cgraph");
    println!("          test      test cgraph\n");
    println!("          clean     clean all the generated files");
    println!("          distclean clean all the generated files and directories");
    println!("          help      commands to this PROgram");
  } else {
    println!("{} is an unsupported command", args[1]);
    println!("use \"{} help\" to know all supported commands", file!());
  }
}

File: script/compile.c
#!/usr/bin/tcc - run

#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define STRING_MAX 1024
#define CMD_MAX    81960
#define FILE_NUM   100
#define STRING_END '\0'

extern char **environ;

int main(int argc, char *argv[])
{
    char *PRO = "cgraph";
    char *_DIR = ".";
    char INC[STRING_MAX], SRC[STRING_MAX], TST[STRING_MAX], LIB[STRING_MAX];

    sprintf(INC, "%s/%s", _DIR, "include");
    sprintf(SRC, "%s/%s", _DIR, "src");
    sprintf(TST, "%s/%s", _DIR, "test");
    sprintf(LIB, "%s/%s", _DIR, "lib");

    char CFLAGS[STRING_MAX];
    char *CC = "cc";
    char *CSFLAGS = "-shared";
    strcpy(CFLAGS, "-std=c89 -Wall -pedantic -fPIC");

    char *mode = "debug";
    if (mode == "debug") {
        strcat(CFLAGS, " -g -DDEBUG");
    } else if (mode == "release") {
        strcat(CFLAGS, " -static -O2");
    }

    // build and clean directories and files
    char *MKDIR = "mkdir";
    char *RM = "rm";
    char *RMFLAGS = "-rf";

    char *RMDIR = "rm";
    char *RMDIRFLAGS = "-rf";

    char *AR = "ar";
    char *ARFLAGS = "-rcs";

    char **env = environ;
    char *path = "PATH";
    for (; NULL != *env; env++) {
        char *senv = *env, *tpath = path;
        for (; STRING_END != *tpath; tpath++, senv++) {
            if (*tpath != *senv) {
                break;
            }
        }
        if (STRING_END == *tpath) {
            fprintf(stdout, "%s\n", *env);
        }
    }

    char CFILES[STRING_MAX][FILE_NUM], OBJ[STRING_MAX];
    DIR *dptr;
    struct dirent *eptr;
    int i = 0, j;
    if (NULL == (dptr = opendir(SRC))) {
        fprintf(stderr, "read directory %s error!\n", SRC);
        exit(-1);
    }
    while (NULL != (eptr = readdir(dptr))) {
        char *file = eptr->d_name;
        for (; STRING_END != *file; file++) {
        }
        if (('c' == *(--file)) && ('.' == *(--file))) {
            sprintf(CFILES[i++], "%s/%s", SRC, eptr->d_name);
        }
    }
    closedir(dptr);

    char CMD[CMD_MAX], BUFFER[CMD_MAX];
    if (argc == 1) {
        sprintf(BUFFER, "%s %s -I%s", CC, CFLAGS, INC);
        for (j = 0; j < i; j++) {
            char *tCFILE = CFILES[j], *tOBJ = OBJ;
            for (; STRING_END != *tCFILE; tCFILE++, tOBJ++) {
                *tOBJ = *tCFILE;
            }
            *tOBJ = STRING_END;
            *(--tOBJ) = 'o';
            sprintf(CMD, "%s -c %s -o %s\n", BUFFER, CFILES[j], OBJ);
            fprintf(stdout, CMD);
            sysetem(CMD);
        }
        // fprintf(stdout, "compile %s\n", LIBSHARED);
        // ${CC} ${CSFLAGS} -o ${LIBSHARED} ${SRC}/*.o
        // fprintf(stdout, "compile %s\n", LIBSTATIC);
        // ${AR} ${ARFLAGS} ${LIBSTATIC} ${SRC}/*.o
    } else if (0 == strcmp("test", argv[1])) {

    } else if (0 == strcmp("clean", argv[1])) {
        sprintf(BUFFER, "%s %s", RM, RMFLAGS);
        for (j = 0; j < i; j++) {
            char *tCFILE = CFILES[j], *tOBJ = OBJ;
            for (; STRING_END != *tCFILE; tCFILE++, tOBJ++) {
                *tOBJ = *tCFILE;
            }
            *tOBJ = STRING_END;
            *(--tOBJ) = 'o';
            sprintf(CMD, "%s %s\n", BUFFER, OBJ);
            fprintf(stdout, "clean %s\n", OBJ);
            system(CMD);
        }
    } else if (0 == strcmp("distclean", argv[1])) {
        sprintf(BUFFER, "%s %s", RMDIR, RMDIRFLAGS);
        for (j = 0; j < i; j++) {
            char *tCFILE = CFILES[j], *tOBJ = OBJ;
            for (; STRING_END != *tCFILE; tCFILE++, tOBJ++) {
                *tOBJ = *tCFILE;
            }
            *tOBJ = STRING_END;
            *(--tOBJ) = 'o';
            sprintf(CMD, "%s %s\n", BUFFER, OBJ);
            fprintf(stdout, "clean %s\n", OBJ);
            system(CMD);
        }
    } else if (0 == strcmp("help", argv[1])) {
        fprintf(stdout, "%s    <target>\n", argv[0]);
        fprintf(stdout, "<target>: \n");
        fprintf(stdout, "                    compile cgraph\n");
        fprintf(stdout, "          test      test cgraph\n");
        fprintf(stdout, "          clean     clean all the generated files\n");
        fprintf(stdout, "          distclean clean all the generated files and "
                        "directories\n");
        fprintf(stdout, "          help      commands to this program\n");
    } else {
        fprintf(stdout, "%s is an unsupported command\n", argv[1]);
        fprintf(stdout, "use \"%s help\" to know all supported commands\n",
                argv[0]);
    }

    return 0;
}